alibi_detect/utils/pytorch/distance.py

```diff
@@ -28,6 +31,70 @@ def squared_pairwise_distance(x: torch.Tensor, y: torch.Tensor, a_min: float = 1
     return dist.clamp_min_(a_min)
 
 
+def batch_compute_kernel_matrix(
+    x: Union[list, np.ndarray, torch.Tensor],
+    y: Union[list, np.ndarray, torch.Tensor],
+    kernel: Union[Callable, nn.Module, nn.Sequential],
+    device: torch.device = None,
+    batch_size: int = int(1e10),
+    preprocess_fn: Callable = None,
+) -> torch.Tensor:
+    """
+    Compute the kernel matrix between x and y by filling in blocks of size
+    batch_size x batch_size at a time.
+
+    Parameters
+    ----------
+    x
+        Reference set.
+    y
+        Test set.
+    kernel
+        PyTorch module.
+    device
+        Device type used. The default None tries to use the GPU and falls back on CPU if needed.
+        Can be specified by passing either torch.device('cuda') or torch.device('cpu').
+    batch_size
+        Batch size used during prediction.
+    preprocess_fn
+        Optional preprocessing function for each batch.
+
+    Returns
+    -------
+    Kernel matrix in the form of a torch tensor
+    """
+    if device is None:
+        device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+    if type(x) != type(y):
+        raise ValueError("x and y should be of the same type")
+
+    if isinstance(x, np.ndarray):
+        x, y = torch.from_numpy(x), torch.from_numpy(y)
+    z = x + y if isinstance(x, list) else torch.cat([x, y], axis=0)
+
+    n_z = len(z)
+    n_batch = int(np.ceil(n_z / batch_size))
+    with torch.no_grad():
+        k_is = []  # type: Union[list, tuple]
+        for i in range(n_batch):
+            istart, istop = i * batch_size, min((i + 1) * batch_size, n_z)
+            i_batch = z[istart:istop]
+            if isinstance(preprocess_fn, Callable):  # type: ignore
+                i_batch = preprocess_fn(i_batch)
+            i_batch = i_batch.to(device)
+            k_ijs = []
+            for j in range(n_batch):
```

Comment:
Might be worth to add a separate github issue to make this function a bit more efficient? E.g. by reusing `i_batch` when `istart, istop == jstart, jstop` and avoiding duplication of the computation for kxy and kyx by identifying beforehand where x stops and y starts?